File: app/src/main/java/net/squanchy/search/view/ItemsAdapter.java
package net.squanchy.search.view;

import java.util.List;
import java.util.Locale;

import net.squanchy.schedule.domain.view.Event;
import net.squanchy.search.SearchResult;
import net.squanchy.search.view.SearchAdapter.ViewTypeId;
import net.squanchy.speaker.domain.view.Speaker;

class ItemsAdapter {

    // These values are "random", we count on them not clashing with the other
    // random values that are used for non-hardcoded numeric IDs (for events
    // and speakers). This is a reasonable assumption in the Long range.
    // In addition, the CRC32 values we use as numeric IDs are always positive.
    private static final long ITEM_ID_EVENTS_HEADER = -100;
    private static final long ITEM_ID_SPEAKERS_HEADER = -101;

    private final SearchResult.Success searchResult;

    ItemsAdapter(SearchResult.Success searchResult) {
        this.searchResult = searchResult;
    }

    int totalItemsCount() {
        if (searchResult.isEmpty()) {
            return 0;
        }

        int totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.getEvents());
        int totalSpeakerItemsCount = totalCountForSectionIncludingHeaders(searchResult.getSpeakers());

        return totalEventItemsCount + totalSpeakerItemsCount;
    }

    boolean isEmpty() {
        return searchResult.isEmpty();
    }

    @ViewTypeId
    int viewTypeAtAbsolutePosition(int position) {
        ensurePositionExists(position);

        int totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.getEvents());

        int adjustedPosition;
        if (position < totalEventItemsCount) {
            adjustedPosition = position;
        } else {
            adjustedPosition = position - totalEventItemsCount;
        }

        if (adjustedPosition == 0) {
            return ViewTypeId.HEADER;
        } else if (position < totalEventItemsCount) {
            return ViewTypeId.EVENT;
        } else {
            return ViewTypeId.SPEAKER;
        }
    }

    long itemIdAtAbsolutePosition(int position) {
        ensurePositionExists(position);

        int totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.getEvents());
        if (totalEventItemsCount > 0 && position < totalEventItemsCount) {
            if (position == 0) {
                return ITEM_ID_EVENTS_HEADER;
            } else {
                return searchResult.getEvents().get(position - 1).getNumericId();
            }
        } else {
            int adjustedPosition = position - totalEventItemsCount;
            // We checked position is in range so it MUST be a speaker
            if (adjustedPosition == 0) {
                return ITEM_ID_SPEAKERS_HEADER;
            } else {
                return searchResult.getSpeakers().get(adjustedPosition - 1).getNumericId();
            }
        }
    }

    Speaker speakerAtAbsolutePosition(int position) {
        ensurePositionExists(position);

        int totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.getEvents());
        if (totalEventItemsCount > 0 && position < totalEventItemsCount) {
            throw new IndexOutOfBoundsException("No speaker at position " + position + ", that is supposed to be in the events sublist");
        }

        int adjustedPosition = position - totalEventItemsCount;
        // We checked position is in range so it MUST be a speaker
        if (adjustedPosition > 0) {
            return searchResult.getSpeakers().get(adjustedPosition - 1);
        } else {
            throw new IndexOutOfBoundsException("No speaker at position " + position + ", that is supposed to be the speakers header");
        }
    }

    Event eventAtAbsolutePosition(int position) {
        ensurePositionExists(position);

        int totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.getEvents());
        if (position == 0) {
            throw new IndexOutOfBoundsException("No event at position " + position + ", that is supposed to be the events header");
        } else if (position - 1 >= totalEventItemsCount) {
            throw new IndexOutOfBoundsException("No event at position " + position + ", that is supposed to be in the speakers sublist");
        }

        return searchResult.getEvents().get(position - 1);
    }

    HeaderType headerTypeAtAbsolutePosition(int position) {
        ensurePositionExists(position);

        int totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.getEvents());
        if (totalEventItemsCount > 0) {
            if (position == 0) {
                return HeaderType.EVENTS;
            } else if (position < totalEventItemsCount) {
                throw new IndexOutOfBoundsException("No header at position " + position + ", that is supposed to be an event");
            }
        }

        int adjustedPosition = position - totalEventItemsCount;
        // We checked position is in range so it MUST be a speaker
        if (adjustedPosition == 0) {
            return HeaderType.SPEAKERS;
        } else {
            throw new IndexOutOfBoundsException("No header at position " + position + ", that is supposed to be a speaker");
        }
    }

    private void ensurePositionExists(int position) {
        int totalItemsCount = totalItemsCount();
        if (position < 0 || position >= totalItemsCount) {
            String message = String.format(Locale.ROOT, "Position %1$d is not valid, must be [0, %2$d)", position, totalItemsCount);
            throw new IndexOutOfBoundsException(message);
        }
    }

    private int totalCountForSectionIncludingHeaders(List sectionItems) {
        int eventsCount = sectionItems.size();
        return eventsCount + headersCountForSectionItemsCount(eventsCount);
    }

    private static int headersCountForSectionItemsCount(int itemsCount) {
        return itemsCount > 0 ? 1 : 0;
    }
}

File: app/src/main/java/net/squanchy/search/view/ItemsAdapter.kt
package net.squanchy.search.view

import java.util.Locale

import net.squanchy.schedule.domain.view.Event
import net.squanchy.search.SearchResult
import net.squanchy.search.view.SearchAdapter.Companion.ViewTypeId
import net.squanchy.speaker.domain.view.Speaker

internal class ItemsAdapter(private val searchResult: SearchResult.Success) {

    val isEmpty: Boolean
        get() = searchResult.isEmpty

    fun totalItemsCount(): Int {
        if (searchResult.isEmpty) {
            return 0
        }

        val totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.events)
        val totalSpeakerItemsCount = totalCountForSectionIncludingHeaders(searchResult.speakers)

        return totalEventItemsCount + totalSpeakerItemsCount
    }

    @ViewTypeId
    fun viewTypeAtAbsolutePosition(position: Int): Int {
        ensurePositionExists(position)

        val totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.events)

        val adjustedPosition: Int
        if (position < totalEventItemsCount) {
            adjustedPosition = position
        } else {
            adjustedPosition = position - totalEventItemsCount
        }

        return when {
            adjustedPosition == 0 -> SearchAdapter.HEADER
            position < totalEventItemsCount -> SearchAdapter.EVENT
            else -> SearchAdapter.SPEAKER
        }
    }

    fun itemIdAtAbsolutePosition(position: Int): Long {
        ensurePositionExists(position)

        val totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.events)
        if (totalEventItemsCount > 0 && position < totalEventItemsCount) {
            return if (position == 0) ITEM_ID_EVENTS_HEADER else searchResult.events[position - 1].numericId
        } else {
            val adjustedPosition = position - totalEventItemsCount
            // We checked position is in range so it MUST be a speaker
            return if (adjustedPosition == 0) ITEM_ID_SPEAKERS_HEADER else searchResult.speakers[adjustedPosition - 1].numericId
        }
    }

    fun speakerAtAbsolutePosition(position: Int): Speaker {
        ensurePositionExists(position)

        val totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.events)
        if (totalEventItemsCount > 0 && position < totalEventItemsCount) {
            throw IndexOutOfBoundsException("No speaker at position $position, that is supposed to be in the events sublist")
        }

        val adjustedPosition = position - totalEventItemsCount
        // We checked position is in range so it MUST be a speaker
        return if (adjustedPosition > 0) {
            searchResult.speakers[adjustedPosition - 1]
        } else {
            throw IndexOutOfBoundsException("No speaker at position $position, that is supposed to be the speakers header")
        }
    }

    fun eventAtAbsolutePosition(position: Int): Event {
        ensurePositionExists(position)

        val totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.events)
        if (position == 0) {
            throw IndexOutOfBoundsException("No event at position $position, that is supposed to be the events header")
        } else if (position - 1 >= totalEventItemsCount) {
            throw IndexOutOfBoundsException("No event at position $position, that is supposed to be in the speakers sublist")
        }

        return searchResult.events[position - 1]
    }

    fun headerTypeAtAbsolutePosition(position: Int): HeaderType {
        ensurePositionExists(position)

        val totalEventItemsCount = totalCountForSectionIncludingHeaders(searchResult.events)
        if (totalEventItemsCount > 0) {
            if (position == 0) {
                return HeaderType.EVENTS
            } else if (position < totalEventItemsCount) {
                throw IndexOutOfBoundsException("No header at position $position, that is supposed to be an event")
            }
        }

        val adjustedPosition = position - totalEventItemsCount
        // We checked position is in range so it MUST be a speaker
        return if (adjustedPosition == 0) {
            HeaderType.SPEAKERS
        } else {
            throw IndexOutOfBoundsException("No header at position $position, that is supposed to be a speaker")
        }
    }

    private fun ensurePositionExists(position: Int) {
        val totalItemsCount = totalItemsCount()
        if (position < 0 || position >= totalItemsCount) {
            val message = String.format(Locale.ROOT, "Position %1\$d is not valid, must be [0, %2\$d)", position, totalItemsCount)
            throw IndexOutOfBoundsException(message)
        }
    }

    private fun totalCountForSectionIncludingHeaders(sectionItems: List<*>): Int {
        val eventsCount = sectionItems.size
        return eventsCount + headersCountForSectionItemsCount(eventsCount)
    }

    companion object {

        // These values are "random", we count on them not clashing with the other
        // random values that are used for non-hardcoded numeric IDs (for events
        // and speakers). This is a reasonable assumption in the Long range.
        // In addition, the CRC32 values we use as numeric IDs are always positive.
        private const val ITEM_ID_EVENTS_HEADER: Long = -100
        private const val ITEM_ID_SPEAKERS_HEADER: Long = -101

        private fun headersCountForSectionItemsCount(itemsCount: Int): Int {
            return if (itemsCount > 0) 1 else 0
        }
    }
}
